### PR TITLE
Trie clone

### PR DESCRIPTION
**Pull request by CoderZhi**
https://github.com/iotexproject/iotex-core/issues/2847

### CODE REVIEW

File: db/trie/mptrie/hashnode.go [hunk 1]
```diff
@@ -8,53 +8,52 @@ package mptrie
 
 type hashNode struct {
 	node
-	mpt     *merklePatriciaTrie
 	hashVal []byte
 }
 
-func newHashNode(mpt *merklePatriciaTrie, ha []byte) *hashNode {
-	return &hashNode{mpt: mpt, hashVal: ha}
+func newHashNode(_ client, ha []byte) *hashNode {
```
Comment (huof6829) [inline db/trie/mptrie/hashnode.go:14]: `_ client` can be deleted

File: db/trie/mptrie/leafnode.go [hunk 2]
```diff
@@ -22,29 +22,29 @@ type leafNode struct {
...
+	if !cli.asyncMode() {
+		if err := l.store(cli); err != nil {
+			return nil, err
+		}
 	}
 	return l, nil
 }
 
-func newLeafNodeFromProtoPb(pb *triepb.LeafPb, mpt *merklePatriciaTrie, hashVal []byte) *leafNode {
+func newLeafNodeFromProtoPb(cli client, pb *triepb.LeafPb, hashVal []byte) *leafNode {
```
Comment (huof6829) [inline db/trie/mptrie/leafnode.go:45]: `cli client` is not used

File: db/trie/mptrie/extensionnode.go [hunk 3]
```diff
@@ -159,52 +159,30 @@ func (e *extensionNode) commonPrefixLength(key []byte) uint8 {
...
-			return hn, nil
-		}
-	}
-	return e, nil
+	return newExtensionNode(cli, path, e.child)
 }
 
-func (e *extensionNode) updateChild(newChild node, hashnode bool) (node, error) {
-	err := e.delete()
+func (e *extensionNode) updateChild(cli client, newChild node, hashnode bool) (node, error) {
```
Comment (huof6829) [inline db/trie/mptrie/extensionnode.go:180]: `hashnode` is not used

File: db/trie/mptrie/merklepatriciatrie.go [hunk 4]
```diff
@@ -222,8 +239,12 @@ func (mpt *merklePatriciaTrie) isEmptyRootHash(h []byte) bool {
...
 func (mpt *merklePatriciaTrie) setRootHash(rootHash []byte) error {
 	if len(rootHash) == 0 || mpt.isEmptyRootHash(rootHash) {
-		emptyRoot := newEmptyRootBranchNode(mpt)
-		return mpt.resetRoot(emptyRoot, mpt.emptyRootHash)
+		emptyRoot, err := newRootBranchNode(mpt, nil, nil, false)
+		if err != nil {
+			return err
+		}
+		mpt.resetRoot(emptyRoot, mpt.emptyRootHash)
+		return nil
```
Comment (huof6829) [inline db/trie/mptrie/merklepatriciatrie.go:247]: return mpt.resetRoot(...)

**Comment by raullenchai** [comment]
Close b/c it has been inactive for 1+yr.

**Pull request closed without merging**